src/main/java/org/reactome/release/dataexport/FTPFileUploader.java

```diff
@@ -0,0 +1,468 @@
+package org.reactome.release.dataexport;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Properties;
+import java.util.stream.Collectors;
+import org.apache.commons.net.ftp.FTPClient;
+import org.apache.commons.net.ftp.FTPFile;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Class for updating Reactome export files on an FTP Server.  Classes which extend this abstract class will provide
+ * methods to:
+ *
+ * 1) Upload new (i.e. current Reactome release) files, created by classes in this project
+ * 2) Delete old (i.e previous Reactome release) files
+ * 3) List files present on the FTP Server for confirmation of successful file upload/deletion.
+ *
+ * @author jweiser
+ */
+public abstract class FTPFileUploader {
+	private static final Logger logger = LogManager.getLogger("mainLog");
+
+	private static final List<String> requiredProperties = Arrays.asList("outputDir", "reactomeVersion");
+
+	private Properties props;
+	private FTPClient ftpClientConnectionToServer;
+
+	/**
+	 * Creates an object for uploading new Reactome files (as well ad deleting old Reactome files) to an FTP Server.
+	 *
+	 * @param props The properties object with the key/value pairs providing connection and path information for
+	 * uploading to the FTP Server
+	 */
+	protected FTPFileUploader(Properties props) {
+		this.props = props;
+		throwIllegalStateExceptionUnlessRequiredPropsPresent();
+	}
+
+	/**
+	 * Returns the logger object for logging info, warnings, errors, etc. from activity in this class and sub-classes
+	 * methods.
+	 *
+	 * @return Logger object for logging class activity
+	 */
+	protected Logger getLogger() {
+		return logger;
+	}
+
+	/**
+	 * Initialization for the FTP Server does the following:
+	 *
+	 * 1) Connects to the FTP Server
+	 * 2) Enters local passive mode
+	 * 3) Logs into the FTP Server with the username and password from getUserName() and getPassword() methods,
+	 * respectively
+	 * 4) Changes to the working directory obtained from the getReactomeDirectoryPathOnFTPServer() method
+	 *
+	 * @throws IOException Thrown if unable, for the FTP Server, to connect, log in, or change working directory
+	 * @see #getUserName()
+	 * @see #getPassword()
+	 * @see #getReactomeDirectoryPathOnFTPServer()
+	 */
+	public void initializeFTPConnectionToServer() throws IOException {
+		getFtpClientToServer().connect(getServerHostName());
+
+		getFtpClientToServer().enterLocalPassiveMode();
+
+		if (loginToFTPServer()) {
+			logger.info("Login successful to " + getServerHostName());
+		} else {
+			logger.error("Login to " + getServerHostName() + " failed");
```

Comment:
Would it be possible to provide additional information here, such as the _cause_ for the failed login? Such as "invalid username/password", "could not find host", etc...

Reply:
I did some checking and the FTPClient API (https://commons.apache.org/proper/commons-net/apidocs/org/apache/commons/net/ftp/FTPClient.html), unfortunately, does not provide a method to get the error message from an unsuccessful command (at least that I can see).

The login can trigger an IOException due to a connection issue, though, which will be proprogated upwards and stop the program.